aic-emu/CmdHandler.cc

```diff
@@ -0,0 +1,572 @@
+
+    SurfaceParams_t surf = bo_data->surf;
+    size_t sizeToRead = surf.width * surf.height * surf.pixelSize;
+    uint8_t* pFrame = new uint8_t[sizeToRead];
+
+    status = GetOneFrame(pFrame, sizeToRead);
+    CHECK_STATUS(status);
+
+    // Copy data to GPU memory indicated by file-descriptor
+    status = m_gfx.Copy(bo_data, pFrame);
+    CHECK_STATUS(status);
+
+    // Maintain fps
+    ManageDisplayReqTime(metadata);
+
+    // Write Data to sockets
+    AIC_LOG();
+    status = WriteIntoSocket(&evInfo->event, sizeof(display_event_t));
+    CHECK_STATUS(status);
+    status = WriteIntoSocket(&evInfo->info, sizeof(buffer_info_t));
+    CHECK_STATUS(status);
+    if (hasCtrl)
+    {
+        auto ctrl = std::static_pointer_cast<display_control_t, void>(payload[1]);
+        status = WriteIntoSocket(ctrl.get(), sizeof(display_control_t));
+        CHECK_STATUS(status);
+    }
+
+    // Clean up and Timing book-keeping
+    if (pFrame)
+        delete[] pFrame;
+
+    m_lastDispReqYmlTS  = metadata->timeStampUs;
+    m_lastDispReqSentTS = duration_cast<microseconds>(system_clock::now().time_since_epoch()).count();
+    if (m_firstDispReqSentTS == 0)
+        m_firstDispReqSentTS = m_lastDispReqSentTS;
+
+    return AICS_ERR_NONE;
+}
+
+
+int CmdHandler::DataExchange(std::vector<std::shared_ptr<void>>& payload, AicEventMetadataPtr& metadata)
+{
+    int status = AICS_ERR_NONE;
+
+    if (payload.empty())
+        return AICS_ERR_PAYLOAD_EMPTY;
+
+    if (metadata->direction == DIRECTION_SEND)
+    {
+        switch(metadata->eventType)
+        {
+        case VHAL_DD_EVENT_DISPINFO_REQ:
+        case VHAL_DD_EVENT_DISPPORT_REQ:
+        {
+            AIC_LOG();
+
+            //send display_event_t
+            auto event = std::static_pointer_cast<display_event_t>(payload[0]);
+            status = WriteIntoSocket(event.get(), sizeof(display_event_t));
+            CHECK_STATUS(status);
+            break;
+        }
+        case VHAL_DD_EVENT_CREATE_BUFFER:
+        {
+            AIC_LOG();
+            //send display_event_t + buffer_info_t + cros_gralloc_handle + Special FDs msg
+            auto evInfo = std::static_pointer_cast<buffer_info_event_t, void>(payload[0]);
+            auto grallocHandle = std::static_pointer_cast<cros_gralloc_handle, void>(payload[1]);
+            m_props = grallocHandle;
+            std::cout << "width = " << m_props->width << " , height = " << m_props->height << std::endl;
+            std::cout << "fds = " << m_props->base.numFds << ", numInts - " << m_props->base.numInts << std::endl;
+
+            SurfaceParams_t surf;
+            m_gfx.DetermineSurfaceParams(surf, m_props->width, m_props->height, m_props->format);
+
+            status = m_gfx.AllocateBo(surf, evInfo->info.remote_handle);
+            CHECK_STATUS((int)status);
+
+            //Write all msgs to socket
+            AIC_LOG();
+            status = WriteIntoSocket(&evInfo->event, sizeof(display_event_t));
```

Comment:
you break this writing in stages... I wonder whether this will/might break anything on a reader side. Basically it depends how reader is implemented... We need to check.

Reply:
The writes are designed to match the reads. There is a write for each read in libvhal HWC. 

I'm thinking this should be fine: pls point out if you think there is a problem.